### ISSUE TITLE
Enhanced versioning reporting updates

### ISSUE THREAD

**Issue by skliper** [nasa/cFE#348]
GitHub repos don't include the name in the tag (since each repo is separate), so the git describe with --match "${NAME}*" comes up blank.  The most recent tag in every branch (babelfish and github) is legit, so no need to filter.

Recommend removing this match, and including the names in the print string:

As in: CFE: %s, OSAL: %s

Looks like MISSION_PSPMODULES is not coming through as expected, suggest adding PSP: %s?

**Comment by skliper** [comment]
_Imported from trac issue 317.  Created by jhageman on 2019-07-25T13:27:41, last modified: 2019-08-29T15:36:38_

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-07-25 13:36:59:_

Removed match and added name in print string: [changeset:ea8a541], branch `trac-317-enhanced-version`

PSP info not fixed yet.

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-07-25 13:45:02:_

@jphickey was there an intent to also git describe the PSP repo?  Is this something you could easily add?

**Comment by skliper** [comment]
_Trac comment by jphickey on 2019-07-29 09:12:28:_

I'm concerned that the lack of a name in the tag is going to cause a project assembled using git subtree to break.  The reason we formatted the tags as `cfe-rel-x.y.z` and `osal-rel-x.y.z` is so that after merging the trees together using a subtree method, the tag names were still relevant and identifiable.  (That is, when using this method you are more or less creating a single repo out of multiple source repos - the tags also get merged and they become ambiguous unless they have a name in them somehow).

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-08-12 18:37:31:_

I'm probably not understanding your use case for subtree tags, so I may be off base (feel free to describe if I got it wrong).

The git subtree tag collision recommendations I've seen are typically along the lines of "don't fetch subtree tags since they often collide".  Seems there is risk in depending on all repos in your project to have unique tags (or branches), since a flight project will likely be pulling from a variety of distributors and the problem would remain even if the cFS framework were to go to extra lengths to ensure unique names.  The current setup is broken as soon as someone subtrees and fetches tags from a non-framework supplied OSAL with a different tagging scheme.  Since we can't force prefixes on all tags of every repo in the cfs ecosystem, I prefer we don't depend on a specific format and instead just report the tag name as-is.

https://www.sourcefield.nl/post/git-subtree-survival-tips/
https://stackoverflow.com/questions/39838718/how-to-pull-all-my-tags-using-the-git-subtree-from-another-repository

-Side note- 
Are you fetching all the subtree tags into the top repo?  If not, when I subtree using the instructions in [https://babelfish.arc.nasa.gov/trac/cfs_test/wiki/TutorialNewMission New Mission] they don't get automatically added.  Seems to me this isn't breaking anything (just don't fetch the subtree tags).  There's also the --no-tags option if it needs to be forced.

I recommend not mixing subtree development activities with distribution management.  It becomes a real headache if mixing branches, crossing repos, etc.

I guessed at use cases but none of them made sense to me.  Checking out a subtree tag replaces your entire distribution with the subtree at that tag.  Diffing against a tag doesn't make sense (since the files are in different locations), etc.  If you want a different version of the subtree in the distribution, subtree pull the tag you want.

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-08-14 14:42:28:_

Reconfirmed following subtree wiki instructions does not fetch tags (git version 2.7.4)

Fetched the tags, and built/ran, output is:

`EVS Port1 42/1/CFE_ES 2: Versions:cFE 6.6.90.0, OSAL 4.2.93.0, PSP 1.3.90.0, chksm 64674`

`EVS Port1 42/1/CFE_ES 91: Mission 2835f43-dirty.sample, cfe-rel-6.6.0a-470-g2835f43-dirty, osal-rel-4.2.1a-674-g2835f43-dirty`

To me this is misleading.  This mission is not using cfe-rel-6.6.0a with 470 changes on top of it, and it's not commit g2835f43.  Similar with osal reporting.

The only useful information I see from this method is the leading part cfe-rel-6.6.0a, and even this is misleading, since really that's just the last tag, and the version information is slightly more helpful (although could be misleading since it's hand edited, and not always incremented).  So I'd consider the current subtree functionality just as broken (or at least no more useful) than reporting the latest git describe from the mission.

My suggestion:
Still report the *_VERSION information (first line, corresponds to individual repo tags) and this is best information available for subtrees (since the rest of the describe is misleading)
Still in cmake do a git describe (with NO match) in base, cfe, osal, and psp locations
In the code, report the mission result and if any of the cfe/osal/psp results are different, report them also (WITH precise text indicating the repo)

That way if it's submoduled or subtreed all the useful information is provided. It's just when subtreed the subtree tags aren't reported (but that information is basically provided by the version line above it).

How's that for an approach that doesn't provide misleading information in either case?

Note it also doesn't require any user reconfiguration.  Could also test for the error case where it isn't a git repository in the location checked, and skip that output also to handle SVN or other CM system.

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-08-16 20:58:48:_

Implemented my suggestion, see [changeset:8fde72f], `trac-317-enhanced-version`

Case where mission directory isn't a git repo, cfe/osal are from babelfish (cFS FW dev style):

`EVS Port1 42/1/CFE_ES 91: Mission sample, CFE: cfe-rel-6.6.0a-48-g1f87707-dirty, OSAL: osal-rel-4.2.1a-49-g05a9f42`

Case where mission directory is subtree (mission distribution style):

`EVS Port1 42/1/CFE_ES 91: Mission cfe-rel-6.6.0a-472-g562f7b2-dirty.sample`

Case where mission directory is submoduled with a local tag in cFE (bundle style):

`EVS Port1 42/1/CFE_ES 91: Mission v6.6.0a-1-gd42368f-dirty.sample, CFE: mytag4.3-dirty, OSAL: osal-rel-4.2.1a-49-g05a9f42`

Case where it's tarballs (no git), or anything besides git (svn, cvs, etc):

`EVS Port1 42/1/CFE_ES 91: Mission sample`


TODO - likely still need to modify test case to get back to full coverage.

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-08-16 21:26:30:_

Marking for CCB review, if accepted will make unit test update it's own ticket (or a general big ticket to scrub all code coverage) that can be done after creating release candidate.  Pending on gcov process getting fixed (or me getting the flags right to do it).

**Comment by skliper** [comment]
_Trac comment by jphickey on 2019-08-19 22:22:37:_

Reviewed.  In general I think the approach and the example output is fine, my only preference is that I would try to do more of the string scrubbing logic at build time (i.e. as part of the cmake script).  Doing this as part of `CFE_ES_TaskInit()` is taking some code space away from other FSW purposes (a little bit, but still, my preference is that if it could be done at compile-time, it probably should be).

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-08-20 09:05:27:_

Building the string in cmake would certainly be preferable.  Would that be something you could implement in the next day or two (along with completing all the other tickets), or should we move forward with this fix and I can write an enhancement ticket targeted for the next build?

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-08-21 13:58:40:_

CCB 8/21/2019 - Reviewed changeset and approved for test.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2019-08-22 16:06:59:_

Merged into branch `ic-ccb-20190821`

**Comment by skliper** [comment]
_Trac comment by jphickey on 2019-08-22 16:09:11:_

This changeset triggers an unused variable warning in the classic build because the EventBuffer and VersionBuffer are only used inside the _ENHANCED_BUILD_ block.

Recommend to defer merging this change set until after the classic build support is dropped.

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-08-22 16:50:10:_

Agreed, move to after classic build drop.

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-08-29 13:33:02:_

CCB 8/28/2019 - Approved for merge to development branch

**Comment by skliper** [comment]
_Trac comment by jphickey on 2019-08-29 15:36:38:_

Merged to `development` branch
